rir/src/compiler/opt/scope_resolution.cpp

```diff
@@ -36,41 +36,55 @@ static bool noReflection(Code* code, Value* callEnv, ScopeAnalysis& analysis,
                 return true;
             bool anyReflection = false;
             res.eachSource([&](ValOrig vo) {
-                if (vo.val->type.maybeLazy())
+                auto v = vo.val->followCastsAndForce();
+                if (v->type.maybeLazy()) {
+                    if (auto ld = LdArg::Cast(v))
+                        if (cls->assumptions().includes(
+                                rir::Assumption::NoReflectiveArgument) ||
+                            cls->assumptions().isEager(ld->id))
+                            return;
                     anyReflection = true;
+                }
             });
             return anyReflection;
         };
 
-        if (auto ld = LdVar::Cast(i)) {
-            auto e = ld->env() == funEnv ? callEnv : ld->env();
-            auto res = analysis.load(state, ld->varName, e);
-            if (anyReflection(res.result))
-                return false;
-            return true;
-        }
+        auto anyReflectionBy = [&](Instruction* i) {
+            if (auto ld = LdVar::Cast(i)) {
+                auto e = ld->env() == funEnv ? callEnv : ld->env();
+                auto res = analysis.load(state, ld->varName, e);
+                if (anyReflection(res.result))
+                    return false;
+                return true;
+            }
 
-        if (auto ld = LdVarSuper::Cast(i)) {
-            auto e = ld->env() == funEnv ? callEnv : ld->env();
-            auto res = analysis.superLoad(state, ld->varName, e);
-            if (anyReflection(res.result))
-                return false;
-            return true;
-        }
+            if (auto ld = LdVarSuper::Cast(i)) {
+                auto e = ld->env() == funEnv ? callEnv : ld->env();
+                auto res = analysis.superLoad(state, ld->varName, e);
+                if (anyReflection(res.result))
+                    return false;
+                return true;
+            }
 
-        if (auto ld = LdFun::Cast(i)) {
-            auto e = ld->env() == funEnv ? callEnv : ld->env();
-            auto res = analysis.loadFun(state, ld->varName, e);
-            if (anyReflection(res.result))
-                return false;
-            return true;
-        }
+            if (auto ld = LdFun::Cast(i)) {
+                auto e = ld->env() == funEnv ? callEnv : ld->env();
+                auto res = analysis.loadFun(state, ld->varName, e);
+                if (anyReflection(res.result))
```

Comment:
Shouldn't all these simiel `if` be the other way around?

Reply:
ugh, that's ugly, also doulbe negation. was not a bug per-se, since the caller also used it the wrong way round :) pushed a cleanup directly to master

Reply:
thanks